Forward merging of trivial-reduction dims in producers

naoyam opened the pull request:
Forwards merging of trivial reductions with non-reduction domains. Mostly just replicated the logic for the broadcast forwarding. Unique changes are marked as PR comments.

TODO
- [x] Clean up tests.

Fixes #1539

Commit:
Forward merging of trivial-reduction dims in producers

naoyam commented on torch/csrc/jit/codegen/cuda/transform_iter.cpp:
```diff
@@ -631,90 +634,72 @@ struct ConsumerForwardingInfo {
+        forwarding_map.emplace(std::make_pair(forwarded_id, merge->out()));
+        compliment_map.emplace(std::make_pair(
+            forwarded_id, std::vector<IterDomain*>{compliment_id}));
+      }
```
Forward trivial reduction merges

naoyam commented on torch/csrc/jit/codegen/cuda/transform_iter.h:
```diff
@@ -159,7 +159,8 @@ class TORCH_CUDA_CU_API ReplayTransformations : public IterVisitor {
   std::unordered_map<IterDomain*, IterDomain*> target2replay_id_map_;
-  std::unordered_map<IterDomain*, IterDomain*> forward_id_map_;
+  std::unordered_map<IterDomain*, IterDomain*> replay_forward_id_map_;
+  std::unordered_map<IterDomain*, IterDomain*> target_forward_id_map_;
```
Separate forwarding maps for targets and replays.

naoyam commented on torch/csrc/jit/codegen/cuda/transform_iter.cpp:
```diff
@@ -261,18 +263,17 @@ BestEffortReplay::BestEffortReplay(
-            ir_utils::filterByType<IterDomain>(replay_expr->inputs());
-        replay_rfactor_ids.insert(inp_ids.begin(), inp_ids.end());
-      }
+    // Only want to forward rfactor in map
+    auto out_ids = ir_utils::filterByType<IterDomain>(replay_expr->outputs());
+    if (std::any_of(out_ids.begin(), out_ids.end(), [](IterDomain* id) {
+          return id->isRFactorProduct();
+        })) {
+      auto inp_ids = ir_utils::filterByType<IterDomain>(replay_expr->inputs());
+      replay_rfactor_ids.insert(inp_ids.begin(), inp_ids.end());
```
Moved out of the innermost loop as it doesn't depend on the loop of `replay_expr->inputs()`.

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -21452,6 +21452,200 @@ TEST_F(NVFuserTest, FusionVectorizeContigIndexWithBroadcast_CUDA) {
+
+  tv3->split(0, 4);
+
+  fusion.printMath();
```
I assume you were just using this test for debugging?

csarofeen replied:
All tests need some print cleanup

csarofeen commented on torch/csrc/jit/codegen/cuda/transform_iter.cpp:
```diff
@@ -631,90 +634,72 @@ struct ConsumerForwardingInfo {
-
-  // If producer has an rfactor root, that's what will match the consumer
-  std::vector<IterDomain*> producer_root = producer->getMaybeRFactorDomain();
+struct ProducerForwardingInfo {
```
Could you add a comment to this similar to the consumer variant?

csarofeen reviewed (approved):
LGTM just some minor cleanup before merging. Thanks!

Commit:
Enable trivial reduction forwarding only when trivial reduction domain
is a root domain.

For example, splitting a reduction domain by 1 and merging it with
another non-reduction domain would result in a trivial-reduction merge.

Probably possible to allow such non-root trivial reduction domains, but that
would mean, e.g., a leaf domain would be mappable yet its root domain
could be unmappable, which seems rather confusing. Considering such
transformations would be unlikely, not enabling forwarding would be fine
and would cause less surprise.

Commit:
test cleanup

Commit:
Add comment

Commit:
Merge branch 'devel' into replay_forward_trivial_reduction

naoyam commented:
@csarofeen I changed this PR to allow forwarding only when trivial-reduction domains are root domains. Non-root trivial reduction domains are those created with split-by-one of reductions domains, for example:

```
tv0: [R0, I1]
tv0->split(0, 1); // the inner output is a trivial reduction domain
tv0->merge(1, 2); // merging with the trivial reduction domain
```

I think forwarding in patterns like this can be confusing. At the root domain level, there's nothing that can be forwarded, so allowing this forwarding would need some more changes in `TransformReplay::replayPasC/CasP`. Unless there's a compelling case where we would need to support forwarding of intermediate trivial reduction domains, I think it makes more sense to make this PR simpler.
